lib/src/over_react_test/react_util.dart

```diff
@@ -481,8 +491,35 @@ Element queryByTestId(dynamic root, String value, {String key = defaultTestIdKey
 ///     </div>
 ///
 ///     queryAllByTestId(renderedInstance, 'value'); // returns both `inner` `<div>`s
-List<Element> queryAllByTestId(dynamic root, String value, {String key = defaultTestIdKey}) {
-  return findDomNode(root).querySelectorAll('[$key~="$value"]');
+List<Element> queryAllByTestId(dynamic root, String value, {String key = defaultTestIdKey, bool searchInShadowDom = false, int shadowDepth}) {
+  return _findDeep(findDomNode(root), _makeTestIdSelector(value, key: key), searchInShadowDom: searchInShadowDom, findMany: true, depth: shadowDepth);
+}
+
+String _makeTestIdSelector(String value, {String key = defaultTestIdKey}) => '[$key~="$value"]';
+
+List<Element> _findDeep(Node root, String itemSelector, {bool searchInShadowDom = false, bool findMany = true, int depth}) {
+  List<Element> nodes = [];
+  final rootQuerySelector = root is ShadowRoot ? root.querySelector : root is Element ? root.querySelector : null;
+  var lightElement = rootQuerySelector(itemSelector);
+  if (!findMany && lightElement != null){
+    return [lightElement];
+  }
+  void recursiveSeek(Node _root, int _currentDepth) {
+    final List<Element> Function(String) rootQuerySelectorAll = _root is ShadowRoot ? _root.querySelectorAll : _root is Element ? _root.querySelectorAll : null;
+    nodes.addAll(rootQuerySelectorAll(itemSelector).toList());
+    if (!findMany && nodes.isNotEmpty) {
+      return;
+    }
+    // This is kinda gross but I figured this was only ever used in test situations so its probably alright.
```

Comment:
Is this comment still necessary? What's only used in test situations?

Reply:
Ah, gotcha.

Back to the comment, I'm not sure specifically what it's trying to call out, and could use some clarification. Did you mean:

```dart
// This method of finding shadow roots may not be optimal, but it's good enough for usage in tests.
```

Reply:
TECHNICALLY... its the only way to find shadow roots so it is actually the MOST optimal it can be without requiring they specify each shadow root host selectors... but kk lol

Reply:
`s/optimal/performant/`

Reply:
Haha that's fine, I just didn't know what exactly you meant by "gross", so I was spitballing a potential explanation.

Reply:
Optimal/performant to me mean the same thing in this context 🤷 so I think we're splitting hairs at this point. It doesn't matter to me what the comment says as long as it's more specific